Core/include/Acts/SpacePointFormation/SpacePointBuilder.hpp

```diff
@@ -0,0 +1,160 @@
+template <typename spacepoint_t>
+class SpacePointBuilder {
+ public:
+  using Measurement = Acts::BoundVariantMeasurement;
+  // Constructor
+  /// @param cfg The configuration for the space point builder
+  /// @param func The function that provides user's SP constructor with global pos, global cov, and sourceLinks.
+  /// @param logger The logging instance
+  SpacePointBuilder(SpacePointBuilderConfig cfg,
+                    std::function<spacepoint_t(
+                        Acts::Vector3, Acts::Vector2,
+                        boost::container::static_vector<const SourceLink*, 2>)>
+                        func,
+                    std::unique_ptr<const Logger> logger =
+                        getDefaultLogger("SpamcePointBuilder", Logging::INFO));
+
+  // Default constructor
+  SpacePointBuilder() = default;
+
+  /// @brief Calculates the space points out of a given collection of measurements
+  /// and stores the results
+  ///
+  /// @param gctx The current geometry context object, e.g. alignment
+  /// @param measurements vector of measurements
+  /// @param spacePointIt Output iterator for the space points
+  template <template <typename...> typename container_t>
+  void calculateSingleHitSpacePoints(
+      const GeometryContext& gctx,
+      const std::vector<const Measurement*>& measurements,
+      std::back_insert_iterator<container_t<spacepoint_t>> spacePointIt) const;
+
+  /// @brief Searches possible combinations of two measurements on different
+  /// surfaces that may come from the same particles
+  ///
+  /// @param gctx The current geometry context object, e.g. alignment
+  /// @param measurementsFront vector of measurements on a surface
+  /// @param measurementsBack vector of measurements on another surface
+  /// @param measurementPairs storage of the measurement pairs
+  void makeMeasurementPairs(
+      const GeometryContext& gctx,
+      const std::vector<const Measurement*>& measurementsFront,
+      const std::vector<const Measurement*>& measurementsBack,
+      std::vector<std::pair<const Measurement*, const Measurement*>>&
+          measurementPairs) const;
+
+  /// @brief Searches possible combinations of two measurements on different
+  /// surfaces that may come from the same particles
+  /// @param gctx The geometry context to use
+  /// @param measurementPair pair of measurements that are space point candidates
+  /// @param stripEndsPair The global positions of the ends of the strip pair
+  /// @param spacePoint The space point to be created.
+  /// @note If no configuration is set, the default values will be used
+  // template <template <typename...> typename container_t>
+  void calculateDoubleHitSpacePoint(
+      const Acts::GeometryContext& gctx,
+      const std::pair<const Measurement*, const Measurement*>& measurementPair,
+      const std::pair<const std::pair<Vector3, Vector3>,
+                      const std::pair<Vector3, Vector3>>& stripEndsPair,
+      spacepoint_t* spacePoint) const;
```

Comment:
I don't understand the ownership of `spacepoint_t*` here.

Reply:
I understand the strip end pairs are enough to do the calculation, and the measurements are for linking, right?

Reply:
Changed it to `shared_ptr`. The measurements are actually needed to calculate the uncertainty.